src/components/molecules/Visualizer/Layers/store.ts

```diff
@@ -68,6 +73,72 @@ export class LayerStore {
     return l;
   }
 
+  #usedIds: string[] = [];
+  #newId = () => {
+    const genResId = () =>
+      "_xxxxxxxxxxxxxxxxxxxxxxxxx".replace(/[x]/g, function () {
+        return ((Math.random() * 16) | 0).toString(16);
+      });
+    let id;
+    do {
+      id = genResId();
+    } while (this.#usedIds.includes(id));
+    this.#usedIds.push(id);
+    return id;
+  };
+
+  #insertLayer = (layerData: AddedLayerData, save = false) => {
+    const tar = layerData.parentId ? this.#map.get(layerData.parentId) : this.#root;
+    if (!tar?.children) return;
+    tar.children.push(layerData.layer);
+    const flattenedLayers = flattenLayers([layerData.layer]);
+    flattenedLayers.forEach(l => {
+      this.#map.set(l.id, l);
+      this.#pmap.set(l.id, this.#pluginLayer(l));
+    });
+    save && !tar.creator && this.#addedLayersData.push(layerData);
+    return layerData.layer.id;
+  };
+
+  #processLayer = (layer: Omit<Layer, "id">): Layer => {
+    const { children, ...self } = layer;
+    self.infobox?.blocks &&
+      self.infobox.blocks.forEach(b => {
+        b.id = this.#newId();
+      });
+    return {
+      ...self,
+      id: this.#newId(),
+      pluginId: "reearth",
+      creator: "pluginAPI",
+      children: children?.map(cl => this.#processLayer(cl)),
+    };
+  };
+
+  add = (layer: Omit<Layer, "id">, parentId?: string) => {
+    const id = this.#insertLayer(
+      {
+        layer: this.#processLayer(layer),
+        parentId,
+      },
+      true,
+    );
+    this.#proot = this.#pluginLayer(this.#root);
+    this.#flattenLayers = flattenLayers(this.#root?.children ?? []);
+    return id;
```

Comment:
```suggestion
  add = (layer: Omit<Layer, "id">, parentId?: string, creator?: string) => {
    const id = this.#insertLayer(
      {
        layer: this.#processLayer(layer, creator),
        parentId,
      },
      true,
    );
    this.#proot = this.#pluginLayer(this.#root);
    this.#flattenLayers = flattenLayers(this.#root?.children ?? []);
    return id;
```

You may want to receive `creator` here.